### ISSUE TITLE
Backtraces stop at thread boundaries

### ISSUE THREAD

**Issue by oli-obk** [rust-lang/miri#4066]
When an action is performed that miri cannot handle (or UB is hit), the backtrace shown ends at the start of the thread in which the failure happened. As an example

```rust
extern {
    fn unknown();
}

fn bad() { unsafe { unknown() } }

fn main() {
    std::thread::spawn(bad);
}
```

fails with

```
error: unsupported operation: can't call foreign function `unknown` on OS `linux`
 --> src/main.rs:5:21
  |
5 | fn bad() { unsafe { unknown() } }
  |                     ^^^^^^^^^ can't call foreign function `unknown` on OS `linux`
  |
  = help: if this is a basic API commonly used on this target, please report an issue with Miri
  = help: however, note that Miri does not aim to support every FFI function out there; for instance, we will not support APIs for things such as GUIs, scripting languages, or databases
  = note: BACKTRACE on thread `unnamed-1`:
  = note: inside `bad` at src/main.rs:5:21: 5:30
```

pointing solely to the function `bad`.

In this simple example that is not an issue, but if a crate spawns a thread somewhere in its internals, it may not be clear why the thread is even running. I don't think it would be expensive to collect a backtrace at every thread spawn and store this information in the thread data structure itself (as we're not going through the backtrace crate, but just iterating over miri's stack).

**Comment by actuallyatoaster** [comment]
This looks interesting, could I give this a go as a first issue?

**Comment by RalfJung** [comment]
Yes, definitely. :)  Please ask on [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/269128-miri) if there are any questions.

**Comment by moabo3li** [comment]
I checked this on current master with the issue example and with a nested-spawn variant.

Current Miri now reports the direct spawn site via:

`note: the current function got called indirectly due to this code`

So the original small example is much better now. However, even with `-Zmiri-backtrace=full`, the diagnostic only shows the child thread's stack plus the direct `std::thread::spawn(...)` span. It does not show the parent thread's stack leading to that spawn, e.g. `main -> library_layer -> spawn_worker`.

Is the remaining intended behavior still to store and print the full spawn-site backtrace, at least in full backtrace mode, or should this issue be considered fixed by the current origin-span note?


I also noticed this may overlap with Priroda's upcoming stack/thread work, but probably only at the data/API layer.

Priroda needs live stack frames for the currently stopped thread, while this issue needs saved parent spawn context from the moment a thread was created. So they are different use cases:

- Priroda `stackTrace`: current live stack
- #4066: saved spawn stack / thread origin

If the remaining #4066 behavior is wanted, would it make sense to represent the spawn context in a small Miri-owned structure/API, so diagnostics can render it now and Priroda could later query it for richer thread views?

**Comment by moabo3li** [comment]
The current diagnostics 

```
error: unsupported operation: can't call foreign function `unknown` on OS `linux`
  --> /tmp/miri-4066-check/thread_backtrace.rs:6:14
   |
 6 |     unsafe { unknown() }
   |              ^^^^^^^^^ unsupported operation occurred here
   |
   = help: this means the program tried to do something Miri does not support; it does not indicate a bug in the program
   = note: this is on thread `unnamed-1`
note: the current function got called indirectly due to this code
  --> /tmp/miri-4066-check/thread_backtrace.rs:10:5
   |
10 |     std::thread::spawn(bad).join().unwrap();
   |     ^^^^^^^^^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error
```

**Comment by RalfJung** [comment]
Indeed, that is the current situation.

@oli-obk do you think this is enough or would you like full backtraces, maybe even multiple threads deep if thread A spawns thread B spawns thread C?

**Comment by oli-obk** [comment]
I think full backtraces would be nice. Once we have it one thread deep, it should be cheap to just copy that and add more on top of it. Though I guess with priroda this extra complexity may not be worth it and we'll just say "if you need more info, use priroda and debug everything with that".
